### DISCUSSION TITLE
NextJS and zustand

### DISCUSSION THREAD

**Discussion by alexander-densley** [pmndrs/zustand#2326] [Q&A] [answered]
In the new nextjs guide, it say:
"No global stores - Because the store should not be shared across requests, it should not be defined as a global variable. Instead, the store should be created per request."

Does this mean we can't use the slices pattern? or do we create a general store provider that handles all the slices? Sorry, just having a hard time wrapping my head around zustand best practices and nextjs. Any advice would be awesome. If anyone has a repo or a tutorial of implementing zustand into a nextjs app with multiple stores or multiple slices that would be incredible, thanks!

**Comment by alexander-densley** [comment]
Ive been trying to implement based off the guide, and Im running into lots of little issues, so a more comprehensive tutorial and an example repo would be incredibly helpful I imagine. Some issues Ive run into are that createContext requires a default value, and useStore is not imported in the guide's code. There are a couple others, but those might be my implementations problem.

**Reply by YenHub** [reply]
Hey @alexander-densley 

Working with Slices in NextJS is not too dissimilar from the normal Store setup.

There are just a couple of tweaks you need to make to break your store down into slices and consume them from a Provider.

The example below largely mimics the suggested approach in the [Zustand Docs](https://docs.pmnd.rs/zustand/guides/nextjs#app-router), albeit; using Slices.

Here is some example boiler-plate to help get you going:

## Tree Structure (NextJS 14)

```
src/
├── app/
│   ├── layout.tsx
│   └── page.tsx
└── lib/
    ├── components/
    │   └── StoreTest/
    │       └── index.tsx
    ├── providers/
    │   └── app-store-provider.tsx
    └── store/
        ├── slices/
        │   └── appSlice.ts
        ├── defaultState.ts
        └── index.ts
```

## Files

<details>

  <summary><strong><code>/src/app/layout.tsx</code></strong></summary>

```tsx
import type { Metadata } from 'next';
import { Inter } from 'next/font/google';

import { AppStoreProvider } from '@/lib/providers/app-store-provider';

const inter = Inter({ subsets: ['latin'] });

export const metadata: Metadata = {
  title: 'Zustand Store Slices Demo',
  description: 'A demonstration of using Slices in NextJS 14 ',
};

export default function RootLayout({
  children,
}: Readonly<{
  children: React.ReactNode;
}>) {
  return (
    <html lang="en">
      <body className={inter.className}>
        <AppStoreProvider>{children}</AppStoreProvider>
      </body>
    </html>
  );
}
```

</details>

<details>

  <summary><strong><code>/src/app/page.tsx</code></strong></summary>

```tsx
import { StoreTest } from '@/lib/components/StoreTest';

export default function Home() {
  return (
    <main>
      <StoreTest />
    </main>
  );
}
```

</details>

<details>

  <summary><strong><code>/src/lib/components/StoreTest/index.tsx</code></strong></summary>

```tsx
'use client';

import { useStore } from '@/lib/providers/app-store-provider';

export const StoreTest: React.FC = () => {
  const { isLoading, setIsLoading } = useStore(state => ({
    isLoading: state.isLoading,
    setIsLoading: state.setIsLoading,
  }));

  return (
    <form action="javascript:void(0);">
      <label htmlFor="storeval">Current Value:</label>
      <br />
      <input
        disabled
        id="storeval"
        name="storeval"
        type="text"
        value={String(isLoading)}
      />
      <br />
      <input
        type="button"
        value="Set Is Loading"
        onClick={() => setIsLoading(!isLoading)}
      />
    </form>
  );
};
```

</details>

<details>

  <summary><strong><code>/src/lib/providers/app-store-provider.tsx</code></strong></summary>

```tsx
'use client';

import { createContext, type ReactNode, useContext, useRef } from 'react';
import { type StoreApi, useStore as useZustandStore } from 'zustand';

import { createAppStore, initStore, Store } from '@/lib/store';

export const AppStoreContext = createContext<StoreApi<Store> | null>(null);

export interface AppStoreProviderProps {
  children: ReactNode;
}

export const AppStoreProvider = ({ children }: AppStoreProviderProps) => {
  const storeRef = useRef<StoreApi<Store>>();
  if (!storeRef.current) storeRef.current = createAppStore(initStore());

  return (
    <AppStoreContext.Provider value={storeRef.current}>
      {children}
    </AppStoreContext.Provider>
  );
};

export const useStore = <T,>(selector: (store: Store) => T): T => {
  const appStoreContext = useContext(AppStoreContext);

  if (!appStoreContext) {
    throw new Error('useStore must be use within AppStoreProvider');
  }

  return useZustandStore(appStoreContext, selector);
};
```

</details>

<details>

  <summary><strong><code>/src/lib/store/slices/appSlice.ts</code></strong></summary>

```ts
import { StateCreator } from 'zustand';

import { Store } from '@/lib/store';

export interface AppState {
  isLoading: boolean;
}

export type AppActions = {
  setIsLoading(isLoading: AppState['isLoading']): void;
  resetAppState(): void;
};

export type AppSlice = AppState & AppActions;

export const initialAppState: AppState = {
  isLoading: false,
};

export const createAppSlice: StateCreator<Store, [], [], AppSlice> = set => ({
  ...initialAppState,
  setIsLoading: isLoading => set({ isLoading }),
  resetAppState: () => set(initialAppState),
});
```

</details>

<details>

  <summary><strong><code>/src/lib/store/defaultState.ts</code></strong></summary>

```ts
import { AppState, initialAppState } from './slices/appSlice';

export type StoreState = AppState;

export const defaultState: AppState = {
  ...initialAppState,
};
```

</details>

<details>

  <summary><strong><code>/src/lib/store/index.ts</code></strong></summary>

```ts
import { createStore } from 'zustand/vanilla';

import { defaultState } from './defaultState';
import { AppSlice, AppState, createAppSlice } from './slices/appSlice';

export type Store = AppSlice;

export const initStore = (): AppState => {
  return defaultState;
};

export const createAppStore = (initState: AppState = defaultState) => {
  return createStore<Store>()((set, get, store) => ({
    ...initState,
    ...createAppSlice(set, get, store),
  }));
};
```

</details>

**Reply by HamishClulee** [reply]
@YenHub You are the real MVP for this... thanks a million!

**Comment by alexander-densley** [comment]
Also, after going through jotai documentation, creating a zustand provider to avoid all this set up would be awesome

**Comment by dbritto-dev** [comment]
@alexander-densley what issues are you facing right now? about slices I don't see why you can't use slice pattern. We are only saying that you should avoid shared store across requests due to that could lead you to some issues even if you don't use Zustand that's something could happen if you are using a global store/variable or module state like Zustand.

**Comment by alexander-densley** [comment]
@dbritto-dev I guess one of the things Im currently confused with is the implementation of these custom providers. Do we have to make a provider for each slice we create, cause if the case thats quite the headache? and if so where and how should they be wrapped around layouts and components. I would love to see a more comprehensive guide explaining whats going on because it is a lot more hands on and boilerplate than traditional zustand

**Reply by dbritto-dev** [reply]
@alexander-densley if you want I have some time today so we can talk about that and may you can help me to improve the docs. Here's the meet -> https://meet.google.com/jvj-ybct-bed

**Reply by tyler-audio** [reply]
@alexander-densley @dbritto-dev 
 I am curious if you guys found a solution or some convention for implementing the slices pattern along with the Next.js integration. I've run into the same issue where I would like to use the slices pattern but cannot find any examples or documentation on how this might look with Next.

**Reply by dbritto-dev** [reply]
@alexander-densley @tyler-audio is the same pattern + context api

**Reply by dbritto-dev** [reply]
I could create a demo if y'all want

**Reply by tyler-audio** [reply]
That would be incredibly helpful and appreciated, I'm sure others would agree.

**Reply by alexander-densley** [reply]
I would love this demo if it's still on the table

**Reply by Vee-ble** [reply]
@dbritto-dev Sir, Can I get reference nextjs + slice pattern? 
Is it possible to apply the slice pattern when implementing initializing-the-store of document next?

**Reply by dbritto-dev** [reply]
@Vee-ble sure

**Reply by alexander-densley** [reply]
any updates on this? no rush, thanks for being so helpful!

**Comment by HasanMothaffar** [comment]
> the store should not be shared across requests, it should not be defined as a global variable.

I'm a bit confused by this statement after reading the Next.js integration guide. Does this only concern server-side (Node.js environment) only or both server and client? I'm using Zustand in my Next.js projects (pages directory) as a global store, but **only on the client-side**, nothing on the server, and from what I've seen each request has its own store. So I was wondering if I should also migrate to using Zustand with React Context.

P.S Thank you for creating/maintaining this awesome library!

**Reply by dbritto-dev** [reply]
@HasanMothaffar let me explain you a little bit, first of all you need to keep in mind that:

- there's no state on the server, just an initial state: the first rule that we need to follow is that we should have the same initial state on server and client if there's something that could be different that lead us to some hydrations errors. I would like to say that we are talking about a build time state, so any initial state should be treated as a build time state so we can ensure that we have the same state on both sides (server and client)
- you can't read/write a state from the server: let's say that since there's no state on the server just an initial state, there's no way to go back and read/write to that state
- everything should be created per request: to keep everything isolated we should create the store per request and then sync the data to make that everything works as expected assuming that there's a store that you read from a global store could lead you to some bugs and hydration errors, for instance going to an authenticated page where we can see the authenticated page first and then the login page, etc

**Reply by HasanMothaffar** [reply]
@dbritto-dev To be honest I didn't really understand your reply. I am familiar with ideas from [2200](https://github.com/pmndrs/zustand/discussions/2200), [2058](https://github.com/pmndrs/zustand/discussions/2058), [182](https://github.com/pmndrs/zustand/issues/182). Maybe I didn't explain my question in a clear way, sorry. I'll use an example to remove ambiguity.

Consider a demo Next.js project (using pages directory).
```
// auth-store.ts
const useAuthStore = create((set) => ({
    user: "Initial",
    setUser: (user: string) => set({ user }),
}));

// pages/first.tsx
const First = () => {
    const [user, setUser] = useAuthStore((s) => [s.user, s.setUser]);
    return (
        <div>
            {user}
            <button onClick={() => setUser("Changed")}>Change User</button>
        </div>
    );
};

// pages/second.tsx
const Second = () => {
    const user = useAuthStore((s) => s.user);
    return <div>{user}</div>;
};
```

As indicated, no special server-side functions are called (such as `getServerSideProps`). This is what I meant by client-only in my question above.

Now if a user performs these actions in order:

| Action | User Value |
| ------------- | ------------- |
| 1. Visit first page | `Initial`  |
| 2. Click on button in first page  | `Changed`  |
| 3. Visit second page from another browser/tab  |  `Initial`  |

It seems like the global store is not shared between requests. This is why I was confused, because the integration docs and all issues/discussions state that the singleton store is shared.

I did another experiment and came to the conclusion that the store is shared between requests only when using server-side functionality (SSR, RSCs).

If I change the above example to this

```
// pages/first.tsx
const First = () => {
    return <></>
};

export const getServerSideProps = () => {
    useAuthStore.setState({ user: "Changed" })
}
```

And after visiting the first page in browser 1, the auth store will have `user: Changed`. Now when visiting second page even from other browsers, a hydration error will occur because on the server, `user = Changed`, but on the client `user = Initial`. In other words, the store is shared between requests.

Now my question is: Is this issue only relevant when I set values from the server side using functions like `getServerSideProps` or using RSCs?

Is it safe for me to use Zustand as a global store when I don't have SSR/RSCs/server-side functionality and store is not initialized with data from server?

Thanks again for your time! I appreciate it.

**Reply by dbritto-dev** [reply, marked as answer]
@HasanMothaffar sure, let me explain you two scenarios:

On nextjs any component is rendered/initialized twice, one on the backend and one the client.

When is safe but:
- not recommended: when you use a global store, the initial state is the same on both (server and client) and the initial state doesn't depends on something on the server, but keep in mind if something changed and now your initial state depends on something on the server, now you are not safe anymore and you need to use the recommended way                               
- recommended: when you use a app/page/route store, the initial state is the same on both (server and client) and the initial state doesn't depend on something on the server

As you can see I'm not talking about RSC or SSR but in those cases we are talking about server, that's why in those cases is common to face hydration issues because people forgot to keep in mind the rules.

By the way try this:

```ts
// auth-store.ts
const useAuthStore = create((set) => ({
    user: `Initial ${Math.random()}`,
    setUser: (user: string) => set({ user }),
}));
```

**Reply by dbritto-dev** [reply]
@HasanMothaffar lmk if that makes sense

**Reply by HasanMothaffar** [reply]
@dbritto-dev Yes, that makes sense to me. I tried the example and faced a hydration error, and I understand the situation in which context is necessary. Thank you!

**Reply by danbeo95** [reply]
hi @dbritto-dev As you said:  "when you use a app/page/route store, the initial state is the same on both (server and client) and the initial state doesn't depend on something on the server". Is it correct with Math.random() case?

**Reply by dbritto-dev** [reply]
@danbeo95 no, is not. We want to make sure that server and client first render are the same, and with Math.random we can't do that.

**Reply by danbeo95** [reply]
@dbritto-dev, thank you for your reply. But I still feel confused. Can you help me with code exp which zustand become the global store. thank a lot

**Reply by dbritto-dev** [reply]
@danbeo95 any time you use a store without context you are using global stores

**Reply by dbritto-dev** [reply]
@danbeo95 context api helps you to isolate a store per component and due to pages are components you can isolate store per pages/requests

**Reply by HasanMothaffar** [reply]
For those coming in late, the short answer to my question is **yes**. You CAN use Zustand as a global store if you're ONLY using Zustand on the client. I had to read almost all discussions to reach this conclusion. Technically, things will work just fine. You are NOT forced to use React Context in this case. The docs somehow imply that there's no possible safe way to use Zustand with Next.js unless using React Context, but I would appreciate it if this point could be made clear.

Here's a quote from the library's maintainer from another discussion:

> but if you ONLY use zustand on the client, a global store is fine.

Link: https://github.com/pmndrs/zustand/discussions/2426#discussioncomment-8860531

That provides a simple answer to my question :)

Now while it's technically correct to use Zustand as a global store ONLY on the client side, it is better to get used to using it with React Context just to make sure no one on the team slips and writes code that causes bugs that are hard to find (for example by using Zustand on the server) (https://github.com/pmndrs/zustand/discussions/2200)

**Reply by KayvT** [reply]
@HasanMothaffar mate, your answer saved me a bunch of headache.. I kept reading and reading and was getting more confused. This helps. I have decided I will still go with the safe options and create contexts for my stores since, in my case, almost all of my stores are used only in ONE route group. 

Thank you!

**Reply by HasanMothaffar** [reply]
You're welcome brother :) I've re-read the docs and honestly they're still kind of ambiguous regarding Next.js integration and React Context.

@dbritto-dev Can we please update the docs and indicate that it's totally OK to use Zustand without React context? Just don't update store in server. There's no need to scare users and tell them about the edge usecase right from the start. Also, there is some misleading info in the discussions where it's stated that it's A MUST to use React Context (e.g. [here](https://github.com/pmndrs/zustand/discussions/2426#discussioncomment-8854979)). There's a pinned discussion [here](https://github.com/pmndrs/zustand/discussions/2740) that discusses this idea from the author of the library.

It would really be helpful to update the docs. You can use my comment above as it's simpler than how the docs are structured now.

**Reply by HasanMothaffar-OSG** [reply]
@dai-shi Can we please update docs regarding this issue? The current docs related to Next.js usage are a bit misleading.

Please refer to my comment here https://github.com/pmndrs/zustand/discussions/2326#discussioncomment-10102892

There's also a discusion here that talks about the same thing here https://github.com/pmndrs/zustand/discussions/2740

**Reply by dai-shi** [reply]
Yes, I would like to add a guide based on #2740 discussion. 
@dbritto-dev is working on 1) doc revamp and 2) new website. We don't need to wait for 2), once 1) is done, it's ready to add it.
Meanwhile, let me add a link to #2740 from the doc.

**Comment by TimurKr** [comment]
Hey @alexander-densley, I have been struggling with this as well but got it working after some time. I created a little library that summarizes all of the boilerplate of working with `slices`, `immer` and `context`. It is literally a single file, so feel free to copy the code and paste it into your app, or install it using npm. I think it might be helpful to many `Next.js`developers, so I'm doing a shameless plug here😅.

[zimmer-context](https://github.com/TimurKr/zimmer-context)

I also thought about making a pull request to add this to the _thir-party-libraries_ section in the docs, but I am not sure it is with it, as it really is about 100 lines of code, what are your thoughts? Maybe part of this could be implemented directly into zustand, to make it easier to create a context provider, and the `useStore` hook, as specified [here](https://docs.pmnd.rs/zustand/guides/nextjs#providing-the-store)

**Reply by dbritto-dev** [reply]
@TimurKr keep in mind that nothing should be global talking about nextjs, everything should isolated at app or page/route level

**Reply by TimurKr** [reply]
Yes, I keep reading this in different discussions and tutorials, but what does that actually mean? 

If I create a store, use context provider and access it only in client components, why wouldn't it work? I can additionally initialize it on the server with data, and then when it is re-rendered on client, everything matches and there is no hydration error. At least that's the way I have it set up in my app and seems to be working so far. 

But I admit I am not really sure what you mean be local/global state. Especially when it comes to useStore and context, which is the recommended way of working in the app directory in next.js.

**Reply by dbritto-dev** [reply]
@TimurKr I explained some parts here -> https://github.com/pmndrs/zustand/discussions/2326#discussioncomment-8516823

**Reply by dbritto-dev** [reply]
yeah, the recommended way to use global stores in general not only Zustand is by creating/initializing them using a context at app or page/route level

**Comment by dalalRohit** [comment]
i am also facing the same issues since last few days. Can't get the head around on how to use multiple store slices along with context-providers & RSC. 

Any example or instructions help would be highly appreciated. Thank you.

**Reply by dbritto-dev** [reply]
@dalalRohit what's the difficult part of implementing slices pattern? should be the same without context, the only difference is that now we are using createStore and Context API but both create and createStore support slices pattern

**Reply by dalalRohit** [reply]
thank you @dbritto-dev for quick reply. 

as per the [next.js setup doc](https://docs.pmnd.rs/zustand/guides/nextjs#app-router), the `layout.tsx` is using `< CounterStoreProvider  />` Provider from `counter-store` slice. What if we have multiple such slices? Does it mean we have to add multiple providers; one for each slice? Isn't that tedious to create multiple such providers? 

Do I do Something like this?

```
<CounterStoreProvider>
   <FishSliceProvider>
       //keep nesting multiple providers here
       {children}
   </FishSliceProvider>
</CounterStoreProvider>

```

Again, you guys are awesome. Thanks for such an amazing lib, cheers! 🍺

**Reply by dbritto-dev** [reply]
@dalalRohit when we talked about slices pattern we are talking about just one store. BTW, counter store is not a slice its a single store. For multiple stores, you need multiple providers.

**Reply by KayvT** [reply]
So this means I can use the slice pattern to have one global store with as many slices as I want, and have only one provider wrapping my entire app. Is my assumption correct? 

Let's say I have the following stores: X.ts, Y.ts, and Z.ts and I decide to make an encapsulating store called wrapper.ts

then I init that wrapper.ts with React Context. 

Is this a safe and a recommended pattern / architecture?

Notes: In my case, the 3 stores are actually all managing the same exact route group. I only broke them down to 3 stores to have better code organization.

**Reply by dbritto-dev** [reply]
yeah, totally valid

**Comment by donprasetiyo** [comment]
Can anyone explain how the tsconfig from the doc (https://docs.pmnd.rs/zustand/guides/nextjs#creating-a-store-per-request) makes the store created per request? Would it be fine if I don't use the same tsconfig?

**Reply by dbritto-dev** [reply]
@donprasetiyo I wonder why you think that creates a store per request, but no. What creates a store per request is the usage of Context API.

**Reply by donprasetiyo** [reply]
Thanks. It was the doc's heading ("Creating a store per request") that made me think that.

**Reply by dbritto-dev** [reply]
@danbeo95 may, you didn't notice that the main title is Creating a store per request and then the other titles are part of that, like h1 and then h2s.
